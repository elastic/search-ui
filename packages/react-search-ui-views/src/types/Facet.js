import PropTypes from "prop-types";
import FacetType from "./FacetType";
import FacetValue from "./FacetValue";

export default PropTypes.shape({
  data: PropTypes.arrayOf(FacetValue).isRequired,
  field: PropTypes.string.isRequired,
  type: FacetType.isRequired
});
