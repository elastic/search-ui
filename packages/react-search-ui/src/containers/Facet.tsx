import React from "react";
import { Component } from "react";
import {
  FacetContainerContext,
  FacetContainerProps,
  FacetViewProps,
  MultiCheckboxFacet
} from "@elastic/react-search-ui-views";
import { helpers } from "@elastic/search-ui";

import { accentFold } from "../helpers";
import { withSearch } from "..";

const { markSelectedFacetValuesFromFilters } = helpers;

type FacetContainerState = {
  searchTerm: string;
  more: number;
};

export class FacetContainer extends Component<
  FacetContainerProps,
  FacetContainerState
> {
  static defaultProps = {
    filterType: "all",
    isFilterable: false,
    show: 5
  };

  constructor(props) {
    super(props);
    this.state = {
      more: props.show,
      searchTerm: ""
    };
  }

  handleClickMore = (totalOptions) => {
    this.setState(({ more }) => {
      let visibleOptionsCount = more + 10;
      const showingAll = visibleOptionsCount >= totalOptions;
      if (showingAll) visibleOptionsCount = totalOptions;

      this.props.a11yNotify("moreFilters", { visibleOptionsCount, showingAll });

      return { more: visibleOptionsCount };
    });
  };

  handleFacetSearch = (searchTerm) => {
    this.setState({ searchTerm });
  };

  render() {
    const { more, searchTerm } = this.state;
    const {
      addFilter,
      className,
      facets,
      field,
      filterType,
      filters,
      label,
      removeFilter,
      setFilter,
      view,
      isFilterable,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      a11yNotify,
      ...rest
    } = this.props;
    const facetsForField = facets[field];

    if (!facetsForField) return null;

    // By using `[0]`, we are currently assuming only 1 facet per field. This will likely be enforced
    // in future version, so instead of an array, there will only be one facet allowed per field.
    const facet = facetsForField[0];

    let facetValues = markSelectedFacetValuesFromFilters(
      facet,
      filters,
      field,
      filterType
    ).data;

    const selectedValues = facetValues
      .filter((fv) => fv.selected)
      .map((fv) => fv.value);

    if (!facetValues.length && !selectedValues.length) return null;

    if (searchTerm.trim()) {
      facetValues = facetValues.filter((option) => {
        let valueToSearch;
        switch (typeof option.value) {
          case "string":
            valueToSearch = accentFold(option.value).toLowerCase();
            break;
          case "number":
            valueToSearch = option.value.toString();
            break;
          case "object":
            valueToSearch =
              typeof option?.value?.name === "string"
                ? accentFold(option.value.name).toLowerCase()
                : "";
            break;

          default:
            valueToSearch = "";
            break;
        }
        return valueToSearch.includes(accentFold(searchTerm).toLowerCase());
      });
    }

    const View: React.ComponentType<FacetViewProps> =
      view || MultiCheckboxFacet;

    const viewProps: FacetViewProps = {
      className,
      label: label,
      onMoreClick: this.handleClickMore.bind(this, facetValues.length),
      onRemove: (value) => {
        removeFilter(field, value, filterType);
      },
      onChange: (value) => {
        setFilter(field, value, filterType);
      },
      onSelect: (value) => {
        addFilter(field, value, filterType);
      },
      options: facetValues.slice(0, more),
      showMore: facetValues.length > more,
      values: selectedValues,
      showSearch: isFilterable,
      onSearch: (value) => {
        this.handleFacetSearch(value);
      },
      searchPlaceholder: `Filter ${label}`,
      ...rest
    };

    return <View {...viewProps} />;
  }
}

export default withSearch<FacetContainerProps, FacetContainerContext>(
  ({ filters, facets, addFilter, removeFilter, setFilter, a11yNotify }) => ({
    filters,
    facets,
    addFilter,
    removeFilter,
    setFilter,
    a11yNotify
  })
)(FacetContainer);
