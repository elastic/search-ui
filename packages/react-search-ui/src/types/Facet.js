import PropTypes from "prop-types";
import FacetType from "./FacetType";
import FacetValue from "./FacetValue";

export default PropTypes.shape({
  data: PropTypes.arrayOf(FacetValue).isRequired,
  // Name of the field this facet is associated with
  field: PropTypes.string.isRequired,
  type: FacetType.isRequired
});
