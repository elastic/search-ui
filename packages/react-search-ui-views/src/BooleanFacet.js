import PropTypes from "prop-types";
import React from "react";

import { FacetValue, FilterValue } from "./types";
import { appendClassName } from "./view-helpers";

function BooleanFacet({
  className,
  label,
  options,
  onChange,
  onRemove,
  values
}) {
  const trueCount = options.find(option => option.value === "true").count;
  const isSelected = values.includes("true");

  const apply = () => onChange("true");
  const remove = () => onRemove("true");
  const toggle = () => {
    isSelected ? remove() : apply();
  };

  return (
    <fieldset className={appendClassName("sui-facet", className)}>
      <legend className="sui-facet__title">{label}</legend>
      <div className="sui-boolean-facet">
        <div className={"sui-boolean-facet__option-input-wrapper"}>
          <label className="sui-boolean-facet__option-label">
            <div className="sui-boolean-facet__option-input-wrapper">
              <input
                className="sui-boolean-facet__checkbox"
                type="checkbox"
                onChange={toggle}
              />
              <span className="sui-boolean-facet__input-text">{label}</span>
            </div>
            <span className="sui-boolean-facet__option-count">{trueCount}</span>
          </label>
        </div>
      </div>
    </fieldset>
  );
}

BooleanFacet.propTypes = {
  className: PropTypes.string,
  label: PropTypes.string.isRequired,
  onRemove: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(FacetValue).isRequired,
  values: PropTypes.arrayOf(FilterValue).isRequired,
  onChange: PropTypes.func.isRequired
};

export default BooleanFacet;
