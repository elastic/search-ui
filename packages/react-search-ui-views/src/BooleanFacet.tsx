import React from "react";
import { FacetViewProps } from "./types";

import { appendClassName } from "./view-helpers";

function BooleanFacet({
  className,
  label,
  options,
  onChange,
  onRemove,
  values
}: FacetViewProps) {
  const trueOptions = options.find((option) => option.value === "true");
  if (!trueOptions) return null;
  const isSelected = values.includes("true");

  const apply = () => onChange("true");
  const remove = () => onRemove("true");
  const toggle = () => {
    isSelected ? remove() : apply();
  };

  return (
    <fieldset className={appendClassName("sui-facet", className)}>
      <legend className="sui-facet__title">{label}</legend>
      <div className="sui-boolean-facet">
        <div className={"sui-boolean-facet__option-input-wrapper"}>
          <label className="sui-boolean-facet__option-label">
            <div className="sui-boolean-facet__option-input-wrapper">
              <input
                data-transaction-name={`facet - ${label}`}
                className="sui-boolean-facet__checkbox"
                type="checkbox"
                onChange={toggle}
                checked={isSelected}
              />
              <span className="sui-boolean-facet__input-text">{label}</span>
            </div>
            <span className="sui-boolean-facet__option-count">
              {trueOptions.count}
            </span>
          </label>
        </div>
      </div>
    </fieldset>
  );
}

export default BooleanFacet;
