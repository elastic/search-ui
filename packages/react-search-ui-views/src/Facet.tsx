import {
  FacetValue,
  FieldValue,
  FilterType,
  SearchContextState
} from "@elastic/search-ui";
import { BaseContainerProps } from "./types";

export type FacetViewProps = {
  className?: string;
  label: string;
  onMoreClick: () => void;
  onRemove: (value: FieldValue) => void;
  onChange: (value: FieldValue) => void;
  onSelect: (value: FieldValue) => void;
  options: FacetValue[];
  showMore: boolean;
  values: any;
  showSearch: boolean;
  onSearch: (value: string) => void;
  searchPlaceholder: string;
};

export type FacetContainerContext = Pick<
  SearchContextState,
  | "addFilter"
  | "removeFilter"
  | "setFilter"
  | "a11yNotify"
  | "filters"
  | "facets"
>;

export type FacetContainerProps = BaseContainerProps & {
  filterType: FilterType;
  show?: number;
  view?: React.ComponentType<FacetViewProps>;
  isFilterable: boolean;
  field: string;
  label: string;
} & FacetContainerContext;
