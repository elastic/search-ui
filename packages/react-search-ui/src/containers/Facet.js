import PropTypes from "prop-types";
import { Component } from "react";
import { MultiCheckboxFacet } from "@elastic/react-search-ui-views";
import { helpers } from "@elastic/search-ui";

import { Facet, Filter, FilterType } from "../types";
import { accentFold } from "../helpers";
import { withSearch } from "..";

const { markSelectedFacetValuesFromFilters } = helpers;

export class FacetContainer extends Component {
  static propTypes = {
    // Props
    className: PropTypes.string,
    field: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    filterType: FilterType,
    show: PropTypes.number,
    view: PropTypes.func,
    isFilterable: PropTypes.bool,
    // State
    filters: PropTypes.arrayOf(Filter).isRequired,
    facets: PropTypes.objectOf(PropTypes.arrayOf(Facet)).isRequired,
    // Actions
    addFilter: PropTypes.func.isRequired,
    removeFilter: PropTypes.func.isRequired,
    setFilter: PropTypes.func.isRequired,
    a11yNotify: PropTypes.func.isRequired
  };

  static defaultProps = {
    filterType: "all",
    isFilterable: false
  };

  constructor({ show = 5 }) {
    super();
    this.state = {
      more: show,
      searchTerm: ""
    };
  }

  handleClickMore = totalOptions => {
    this.setState(({ more }) => {
      let visibleOptionsCount = more + 10;
      const showingAll = visibleOptionsCount >= totalOptions;
      if (showingAll) visibleOptionsCount = totalOptions;

      this.props.a11yNotify("moreFilters", { visibleOptionsCount, showingAll });

      return { more: visibleOptionsCount };
    });
  };

  handleFacetSearch = searchTerm => {
    this.setState({ searchTerm });
  };

  render() {
    const { more, searchTerm } = this.state;
    const {
      addFilter,
      className,
      facets,
      field,
      filterType,
      filters,
      label,
      removeFilter,
      setFilter,
      view,
      isFilterable,
      // eslint-disable-next-line no-unused-vars
      a11yNotify,
      ...rest
    } = this.props;
    const facetsForField = facets[field];

    if (!facetsForField) return null;

    // By using `[0]`, we are currently assuming only 1 facet per field. This will likely be enforced
    // in future version, so instead of an array, there will only be one facet allowed per field.
    const facet = facetsForField[0];

    let facetValues = markSelectedFacetValuesFromFilters(
      facet,
      filters,
      field,
      filterType
    ).data;

    const selectedValues = facetValues
      .filter(fv => fv.selected)
      .map(fv => fv.value);

    if (!facetValues.length && !selectedValues.length) return null;

    if (searchTerm.trim()) {
      facetValues = facetValues.filter(option =>
        accentFold(option.value)
          .toLowerCase()
          .includes(accentFold(searchTerm).toLowerCase())
      );
    }

    const View = view || MultiCheckboxFacet;

    return View({
      className,
      label: label,
      onMoreClick: this.handleClickMore.bind(this, facetValues.length),
      onRemove: value => {
        removeFilter(field, value, filterType);
      },
      onChange: value => {
        setFilter(field, value, filterType);
      },
      onSelect: value => {
        addFilter(field, value, filterType);
      },
      options: facetValues.slice(0, more),
      showMore: facetValues.length > more,
      values: selectedValues,
      showSearch: isFilterable,
      onSearch: value => {
        this.handleFacetSearch(value);
      },
      searchPlaceholder: `Filter ${field}`,
      ...rest
    });
  }
}

export default withSearch(
  ({ filters, facets, addFilter, removeFilter, setFilter, a11yNotify }) => ({
    filters,
    facets,
    addFilter,
    removeFilter,
    setFilter,
    a11yNotify
  })
)(FacetContainer);
