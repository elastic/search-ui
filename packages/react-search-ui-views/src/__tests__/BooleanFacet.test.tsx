import React from "react";
import BooleanFacet from "../BooleanFacet";
import { shallow, mount } from "enzyme";
import { FacetViewProps } from "../Facet";

const params: FacetViewProps = {
  label: "A Facet",
  onRemove: jest.fn(),
  onChange: jest.fn(),
  onMoreClick: jest.fn(),
  onSearch: jest.fn(),
  onSelect: jest.fn(),
  searchPlaceholder: "Search",
  showMore: false,
  showSearch: false,
  options: [
    {
      value: "true",
      count: 10
    }
  ],
  values: []
};

it("renders", () => {
  const wrapper = shallow(<BooleanFacet {...params} />);
  expect(wrapper).toMatchSnapshot();
});

it("renders with className prop applied", () => {
  const customClassName = "test-class";
  const wrapper = shallow(
    <BooleanFacet className={customClassName} {...params} />
  );
  const { className } = wrapper.props();
  expect(className.includes(customClassName)).toBe(true);
});

it("onChange is called on click", () => {
  const wrapper = mount(<BooleanFacet {...params} />);

  wrapper.find("input").simulate("change");
  expect(params.onChange).toHaveBeenCalledTimes(1);
});

it("onRemove is called on click", () => {
  const wrapper = mount(
    <BooleanFacet
      {...{
        ...params,
        values: ["true"]
      }}
    />
  );

  wrapper.find("input").simulate("change");
  expect(params.onRemove).toHaveBeenCalledTimes(1);
});

it("will not render when there are no true options", () => {
  params.options = [];
  const wrapper = shallow(<BooleanFacet {...params} />);
  expect(wrapper).toMatchSnapshot();
});
