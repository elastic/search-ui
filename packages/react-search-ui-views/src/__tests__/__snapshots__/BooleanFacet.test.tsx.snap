// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<fieldset
  className="sui-facet"
>
  <legend
    className="sui-facet__title"
  >
    A Facet
  </legend>
  <div
    className="sui-boolean-facet"
  >
    <div
      className="sui-boolean-facet__option-input-wrapper"
    >
      <label
        className="sui-boolean-facet__option-label"
      >
        <div
          className="sui-boolean-facet__option-input-wrapper"
        >
          <input
            checked={false}
            className="sui-boolean-facet__checkbox"
            onChange={[Function]}
            type="checkbox"
          />
          <span
            className="sui-boolean-facet__input-text"
          >
            A Facet
          </span>
        </div>
        <span
          className="sui-boolean-facet__option-count"
        >
          10
        </span>
      </label>
    </div>
  </div>
</fieldset>
`;

exports[`will not render when there are no true options 1`] = `""`;
